impl/src/main/java/org/jboss/weld/event/ObserverNotifier.java

```diff
@@ -347,9 +354,35 @@ public RuntimeException apply(Type eventType) {
             }), executor);
         }
 
-        // TODO if timeout set then wrap the completableFuture
+        // If NotificationOptionKeys.TIMEOUT is set, we will trigger the counter and use CompletableFuture.anyOf()
+        if (options.get(NotificationOptionKeys.TIMEOUT) != null) {
+            // TODO - any beter way to do the cast?
+            return new AsyncEventDeliveryStage<>((CompletableFuture<U>)CompletableFuture.anyOf(completableFuture, initiateTimer(options)), executor);
+        } else {
+            return new AsyncEventDeliveryStage<>(completableFuture, executor);
+        }
+    }
 
-        return new AsyncEventDeliveryStage<>(completableFuture, executor);
+    /**
+     * Checks if the ScheduledExecutorServices was provided.
+     * Verifies the input value for timeout.
+     * Starts the timer thread and returns it as CompletableFuture.
+     */
+    private <T> CompletableFuture<T> initiateTimer(NotificationOptions options) {
+        // throw exception if we don't have scheduled executor available
+        if (timerExecutor == null) {
```

Comment:
I would move this check right after `if (options.get(NotificationOptionKeys.TIMEOUT) != null) {`